source/common/upstream/upstream_impl.h

```diff
@@ -964,7 +964,17 @@ class ClusterInfoImpl : public ClusterInfo,
     return http_protocol_options_->alternate_protocol_cache_options_;
   }
 
-  absl::optional<std::string> edsServiceName() const override { return eds_service_name_; }
+  absl::optional<std::string> edsServiceName() const override {
+    if (eds_service_name_ != nullptr) {
+      return *eds_service_name_;
+    }
+    // If `eds_service_name` is not set, but it's an EDS cluster, use cluster
+    // name.
+    if (type_ == envoy::config::cluster::v3::Cluster::EDS) {
+      return name_;
+    }
```

Comment:
While this may be correct, it seems to break the previous behavior (IIUC it returns nullopt in this case). This change may be ok, just wondering if it is warrented.

Reply:
This allows us not to store `eds_name`, when it's the same as `name`, or when it's empty. Since `EdsClusterImpl` falls back on `name` in such cases already, I think this is a reasonable improvement.

Reply:
Ok, hopefully it won't break anyone.
It should be noted that this has visible impact on clusters-dump in admin (see [here](https://github.com/envoyproxy/envoy/blob/4f2828650d7136e735f8f54e7fb0c9f74ad90ae7/source/server/admin/clusters_handler.cc#L118-L121) and [here](https://github.com/envoyproxy/envoy/blob/4f2828650d7136e735f8f54e7fb0c9f74ad90ae7/source/server/admin/clusters_handler.cc#L219-L222)), and on [load-stats report](https://github.com/envoyproxy/envoy/blob/4f2828650d7136e735f8f54e7fb0c9f74ad90ae7/source/common/upstream/load_stats_reporter.cc#L72-L74). So a release note should be added to this change.

Also please update the [unit tests](https://github.com/envoyproxy/envoy/blob/main/test/common/upstream/upstream_impl_test.cc) to cover the new behavior.

Reply:
OK, if it has visible impact (I didn't realize that was the case), then it's not worth the trouble, I think. Especially since `EdsClusterImpl` falls back on cluster name anyway.

Reply:
Restored original behavior.